### PR TITLE
AELO: plot hazard curves, disaggregation by sources and governing MCE

### PR DESCRIPTION
**Pull request by ptormene**
Supersedes https://github.com/gem/oq-engine/pull/9072, part of https://github.com/gem/oq-engine/issues/9043.

### CODE REVIEW

File: openquake/calculators/postproc/compute_rtgm.py [hunk 1]
```diff
@@ -228,22 +239,25 @@ def get_mce_asce7(prob_mce, det_imt, DLLs, dstore):
...
-            
+
     det_mce = {}
     mce = {}  # imt -> MCE
     prob_mce_out = {}
     for i, imt in enumerate(det_imt):
         det_mce[imt] = max(det_imt[imt], DLLs[i])
-        mce[imt] = min(prob_mce[i], det_mce[imt]) 
+        mce[imt] = min(prob_mce[i], det_mce[imt])
+        # FIXME
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:249]: Why this comment? Is it relevant or it can be removed?

Comment (kejohnso) [reply openquake/calculators/postproc/compute_rtgm.py:249]: I removed it

File: openquake/calculators/postproc/compute_rtgm.py [hunk 2]
```diff
@@ -326,6 +343,285 @@ def get_asce41(dstore, mce, facts):
...
+        imtls_dict[imt] = imls
+    # get info : specific to disagg by src
+    df = dstore['mean_rates_by_src'].to_dframe().set_index('src_id')
+    grouped_m = df.groupby(['src_id', 'site_id', 'imt']).agg(
+        {"value": lambda x: list(x)}).reset_index()
+    # remove the sources that aren't contributing at all to the hazard
+    mask = grouped_m.value.apply(lambda x: sum(x) > 0)
+    gm = grouped_m[mask].reset_index()
+    grouped_2 = gm.groupby(['imt', 'src_id']).agg(
+        {"value": lambda x: np.array(x)}).reset_index()
```
[the remark this answers is not in the document]
Comment (kejohnso) [reply openquake/calculators/postproc/compute_rtgm.py]: okay this works, updated

File: openquake/calculators/postproc/compute_rtgm.py [hunk 3]
```diff
@@ -326,6 +340,284 @@ def get_asce41(dstore, mce, facts):
...
+                          imt_list):
+    plt.figure(figsize=(12, 9))
+    plt.rcParams.update({'font.size': 16})
+    colors = mpl.colormaps['viridis'].reversed()._resample(3)
+    patterns = ['-', '--', ':']
+    for i, imt in enumerate(imt_list):
+        lab = _get_label(imt)
+        plt.loglog(imls[i], AFE[i], color=colors(i), linestyle=patterns[i],
+                   label=lab, linewidth=3, zorder=1)
+        if i == 2:
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:359]: This is not readable at all. We should compare with the IMT corresponding to `i==2`

File: openquake/calculators/postproc/compute_rtgm.py [hunk 4]
```diff
@@ -326,6 +340,284 @@ def get_asce41(dstore, mce, facts):
...
+
+def disaggr_by_src(dstore, imtls):
+    imtls_dict = {}
+    for imt, imls in imtls.items():
+        imls = [iml for iml in imls]
+        imtls_dict[imt] = imls
+    # get info : specific to disagg by src
+    df = dstore['mean_rates_by_src'].to_dframe().set_index('src_id')
+    grouped_m = df.groupby(['src_id', 'site_id', 'imt']).agg(
+        {"value": lambda x: list(x)}).reset_index()
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:401]: `lambda x: list(x)` can be replaced with `list`

File: openquake/calculators/postproc/compute_rtgm.py [hunk 5]
```diff
@@ -326,6 +340,284 @@ def get_asce41(dstore, mce, facts):
...
+    # remove the sources that aren't contributing at all to the hazard
+    mask = grouped_m.value.apply(lambda x: sum(x) > 0)
+    gm = grouped_m[mask].reset_index()
+    grouped_2 = gm.groupby(['imt', 'src_id']).agg(
+        {"value": np.array}).reset_index()
+    total_poe = []
+    for wp in grouped_2.value.values:
+        wsp = []
+        if 'list' in str(type(wp)):
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:410]: Please use `isinstance(wp, list)`

File: openquake/calculators/postproc/compute_rtgm.py [hunk 6]
```diff
@@ -326,6 +340,284 @@ def get_asce41(dstore, mce, facts):
...
+
+        # identify contribution of largest contributor, make color scale
+        largest_contr = np.max(out_contr_all)
+        sample = sum(out_contr_all > fact*largest_contr)
+        viridis = mpl.colormaps['viridis'].reversed()._resample(sample)
+
+        # find and plot the sources, highlighting the ones that contribute more
+        # than 10% of largest contributor
+
+        i = j = 0
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:479]: What's the meaning of `i` and `j`? Can we have better names or at least a comment?

File: openquake/calculators/postproc/compute_rtgm.py [hunk 7]
```diff
@@ -326,6 +340,284 @@ def get_asce41(dstore, mce, facts):
...
+    dinfo = get_info(dstore)
+    # site is always 0 for a single-site calculation
+    # get imls and imts, make arrays
+    imtls = dinfo['imtls']
+    # separate imts and imls
+    AFE, afe_target, imls = [], [], []
+    # NOTE: for meanHCs_afe_RTGM and disaggr_by_src we want to display these
+    # three imts, that are mandatory in this context. For the plot of governing
+    # MCE we read imts from the imtls
+    imt_list = ['PGA', 'SA(0.2)', 'SA(1.0)']
```
Comment (micheles) [inline openquake/calculators/postproc/compute_rtgm.py:586]: We already have the constant `imts`

File: openquake/calculators/postproc/compute_rtgm.py (hunk 4 above)
Comment (kejohnso) [reply openquake/calculators/postproc/compute_rtgm.py:401]: replaced

File: openquake/calculators/postproc/compute_rtgm.py (hunk 5 above)
Comment (kejohnso) [reply openquake/calculators/postproc/compute_rtgm.py:410]: changed

File: openquake/calculators/postproc/compute_rtgm.py (hunk 7 above)
Comment (kejohnso) [reply openquake/calculators/postproc/compute_rtgm.py:586]: replaced `imt_list` with `imts` except for in `plot_governing_mce`

**Comment by kejohnso** [comment]
As of [197467](https://github.com/gem/oq-engine/pull/9089/commits/19746700b1b4f4ef0457aab181db9aa54c6b471a), this exports ASCE7 and 41 even when the hazard is low.

**Review by micheles — APPROVED** [review]

**Pull request closed without merging**